Import of insights webtests fail

aztfy v0.3.0(b1fe27d)
Terraform v1.1.9
on darwin_amd64
 provider registry.terraform.io/hashicorp/azurerm v3.5.0

Import in batch mode:
```
aztfy -b -f -k -m ../mapping.file my-rg
```
mapping.file
```
{
    "/subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webtests/testing my-webpage": "azurerm_application_insights_web_test.my_webpage"
}
```
Output:
```
[ERROR] Failed to import /subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webtests/testing my-webpage as azurerm_application_insights_web_test.my_webpage: exit status 1

Error: parsing Resource ID "/subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webtests/testing my-webpage": ID was missing the `webTests` element
```
It looks like there is confusion if it is webtests vs webTests

new_mapping.file
```
{
    "/subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webTests/testing my-webpage": "azurerm_application_insights_web_test.my_webpage"
}
```
Output:
```
[WARN] No mapping information for resource: /subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webtests/testing my-webpage as azurerm_application_insights_web_test.my_webpage, skip it
```

I am not sure if this is aztfy eller terraform/azurerm.

terraform import works with webTests, but not with webtests:
```
terraform import azurerm_application_insights_web_test.my_webpage '/subscriptions/000000-0000-0000-0000-000000000000/resourceGroups/my-rg/providers/microsoft.insights/webTests/testing my-webpage'
```

--- a/internal/armtemplate/armtemplate.go
+++ b/internal/armtemplate/armtemplate.go
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"regexp"
 	"strings"
+
+	"github.com/Azure/aztfy/internal/client"
 )
 
 type Template struct {
@@ -24,7 +26,52 @@ type ResourceId struct {
 
 var ResourceGroupId = ResourceId{}
 
-func NewResourceIdFromCallExpr(expr string) (*ResourceId, error) {
+func ParseResourceId(id string) (*ResourceId, error) {
+	id = strings.TrimPrefix(id, "/")
+	id = strings.TrimSuffix(id, "/")
+	segs := strings.Split(id, "/")
+	if len(segs)%2 != 0 {
+		return nil, fmt.Errorf("invalid resource id format of %q: amount of segments is not even", id)
+	}
+	// ==4: resource group
+	// >=8: general resources resides in a resource group
+	if len(segs) != 4 && len(segs) < 8 {
+		return nil, fmt.Errorf("invalid resource id format of %q: amount of segments is too small", id)
+	}
+	if segs[0] != "subscriptions" {
+		return nil, fmt.Errorf("invalid resource id format of %q: the 1st segment is not subscriptions", id)
+	}
+	segs = segs[2:]
+	if !strings.EqualFold(segs[0], "resourcegroups") {
+		return nil, fmt.Errorf("invalid resource id format of %q: the 2nd segment is not resourcegroups (case insensitive)", id)
+	}
+	segs = segs[2:]
+
+	if len(segs) == 0 {
+		return &ResourceGroupId, nil
+	}
+
+	if segs[0] != "providers" {
+		return nil, fmt.Errorf("invalid resource id format of %q: the 3rd segment is not providers", id)
+	}
+	providerName := segs[1]
+	segs = segs[2:]
+
+	t := []string{providerName}
+	n := []string{}
+
+	for i := 0; i < len(segs); i += 2 {
+		t = append(t, segs[i])
+		n = append(n, segs[i+1])
+	}
+
+	return &ResourceId{
+		Type: strings.Join(t, "/"),
+		Name: strings.Join(n, "/"),
+	}, nil
+}
+
+func ParseResourceIdFromCallExpr(expr string) (*ResourceId, error) {
 	matches := regexp.MustCompile(`^\[resourceId\(([^,]+), (.+)\)]$`).FindAllStringSubmatch(expr, 1)
 	if len(matches) == 0 {
 		return nil, fmt.Errorf("the resourceId call expression %q is not valid (no match)", expr)
@@ -50,23 +97,24 @@ func NewResourceIdFromCallExpr(expr string) (*ResourceId, error) {
 	}, nil
 }
 
-// ID returns the azure resource id
+// ID converts the ARM ResourceId to its ARM resource ID literal, based on the specified subscription id and resource
+// group name.
 func (res ResourceId) ID(sub, rg string) string {
 	typeSegs := strings.Split(res.Type, "/")
 	nameSegs := strings.Split(res.Name, "/")
 
-	out := []string{"/subscriptions", sub, "resourceGroups", rg}
+	segs := []string{"/subscriptions", sub, "resourceGroups", rg}
 	if len(typeSegs) != 1 {
 		if len(typeSegs)-1 != len(nameSegs) {
 			panic(fmt.Sprintf("The resource of type %q and name %q is not a valid identifier", res.Type, res.Name))
 		}
-		out = append(out, "providers", typeSegs[0])
+		segs = append(segs, "providers", typeSegs[0])
 		for i := 0; i < len(nameSegs); i++ {
-			out = append(out, typeSegs[i+1])
-			out = append(out, nameSegs[i])
+			segs = append(segs, typeSegs[i+1])
+			segs = append(segs, nameSegs[i])
 		}
 	}
-	return strings.Join(out, "/")
+	return strings.Join(segs, "/")
 }
 
 type ResourceIds []ResourceId
@@ -79,7 +127,7 @@ func (resids *ResourceIds) UnmarshalJSON(b []byte) error {
 
 	var ids ResourceIds
 	for _, residExpr := range residExprs {
-		id, err := NewResourceIdFromCallExpr(residExpr)
+		id, err := ParseResourceIdFromCallExpr(residExpr)
 		if err != nil {
 			return err
 		}
@@ -101,32 +149,43 @@ type FQResource struct {
 	DependsOn  []string
 }
 
-func (tpl FQTemplate) DependencyInfo() map[string][]string {
+func (tpl FQTemplate) DependencyInfo() (map[string][]string, error) {
 	s := map[string][]string{}
 	for _, res := range tpl.Resources {
 		if len(res.DependsOn) == 0 {
-			s[res.Id] = []string{ResourceGroupId.ID(tpl.subId, tpl.rg)}
+			rgid, _ := ResourceGroupId.ProviderId(tpl.subId, tpl.rg, nil)
+			s[res.Id] = []string{rgid}
 			continue
 		}
 		s[res.Id] = res.DependsOn
 	}
-	return s
+	return s, nil
 }
 
-func (tpl Template) Qualify(subId, rg string) FQTemplate {
+// Qualify converts the raw ARM template to a fully qualified form in turns of the resource ids, where the ids are converted from the
+// ARM resource id expression to the corresponding TF resource id form.
+func (tpl Template) Qualify(subId, rg string, b *client.ClientBuilder) (*FQTemplate, error) {
 	fqtpl := FQTemplate{
 		subId: subId,
 		rg:    rg,
 	}
 	for _, res := range tpl.Resources {
+		id, err := res.ResourceId.ProviderId(subId, rg, b)
+		if err != nil {
+			return nil, err
+		}
 		fqres := FQResource{
-			Id:         res.ResourceId.ID(subId, rg),
+			Id:         id,
 			Properties: res.Properties,
 		}
 		for _, d := range res.DependsOn {
-			fqres.DependsOn = append(fqres.DependsOn, d.ID(subId, rg))
+			id, err := d.ProviderId(subId, rg, b)
+			if err != nil {
+				return nil, err
+			}
+			fqres.DependsOn = append(fqres.DependsOn, id)
 		}
 		fqtpl.Resources = append(fqtpl.Resources, fqres)
 	}
-	return fqtpl
+	return &fqtpl, nil
 }

--- a/internal/armtemplate/armtemplate.go
+++ b/internal/armtemplate/armtemplate.go
@@ -24,51 +24,6 @@ type ResourceId struct {
 
 var ResourceGroupId = ResourceId{}
 
-func NewResourceId(id string) (*ResourceId, error) {
-	id = strings.TrimPrefix(id, "/")
-	id = strings.TrimSuffix(id, "/")
-	segs := strings.Split(id, "/")
-	if len(segs)%2 != 0 {
-		return nil, fmt.Errorf("invalid resource id format of %q: amount of segments is not even", id)
-	}
-	// ==4: resource group
-	// >=8: general resources resides in a resource group
-	if len(segs) != 4 && len(segs) < 8 {
-		return nil, fmt.Errorf("invalid resource id format of %q: amount of segments is too small", id)
-	}
-	if segs[0] != "subscriptions" {
-		return nil, fmt.Errorf("invalid resource id format of %q: the 1st segment is not subscriptions", id)
-	}
-	segs = segs[2:]
-	if !strings.EqualFold(segs[0], "resourcegroups") {
-		return nil, fmt.Errorf("invalid resource id format of %q: the 2nd segment is not resourcegroups (case insensitive)", id)
-	}
-	segs = segs[2:]
-
-	if len(segs) == 0 {
-		return &ResourceGroupId, nil
-	}
-
-	if segs[0] != "providers" {
-		return nil, fmt.Errorf("invalid resource id format of %q: the 3rd segment is not providers", id)
-	}
-	providerName := segs[1]
-	segs = segs[2:]
-
-	t := []string{providerName}
-	n := []string{}
-
-	for i := 0; i < len(segs); i += 2 {
-		t = append(t, segs[i])
-		n = append(n, segs[i+1])
-	}
-
-	return &ResourceId{
-		Type: strings.Join(t, "/"),
-		Name: strings.Join(n, "/"),
-	}, nil
-}
-
 func NewResourceIdFromCallExpr(expr string) (*ResourceId, error) {
 	matches := regexp.MustCompile(`^\[resourceId\(([^,]+), (.+)\)]$`).FindAllStringSubmatch(expr, 1)
 	if len(matches) == 0 {
@@ -134,17 +89,44 @@ func (resids *ResourceIds) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
-type DependencyInfo map[ResourceId][]ResourceId
+type FQTemplate struct {
+	subId     string
+	rg        string
+	Resources []FQResource
+}
 
-func (tpl Template) DependencyInfo() DependencyInfo {
-	s := map[ResourceId][]ResourceId{}
+type FQResource struct {
+	Id         string
+	Properties interface{}
+	DependsOn  []string
+}
+
+func (tpl FQTemplate) DependencyInfo() map[string][]string {
+	s := map[string][]string{}
 	for _, res := range tpl.Resources {
 		if len(res.DependsOn) == 0 {
-			s[res.ResourceId] = []ResourceId{ResourceGroupId}
+			s[res.Id] = []string{ResourceGroupId.ID(tpl.subId, tpl.rg)}
 			continue
 		}
-
-		s[res.ResourceId] = res.DependsOn
+		s[res.Id] = res.DependsOn
 	}
 	return s
+}
+
+func (tpl Template) Qualify(subId, rg string) FQTemplate {
+	fqtpl := FQTemplate{
+		subId: subId,
+		rg:    rg,
+	}
+	for _, res := range tpl.Resources {
+		fqres := FQResource{
+			Id:         res.ResourceId.ID(subId, rg),
+			Properties: res.Properties,
+		}
+		for _, d := range res.DependsOn {
+			fqres.DependsOn = append(fqres.DependsOn, d.ID(subId, rg))
+		}
+		fqtpl.Resources = append(fqtpl.Resources, fqres)
+	}
+	return fqtpl
 }

--- a/internal/armtemplate/armtemplate_hack.go
+++ b/internal/armtemplate/armtemplate_hack.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"strings"
 
+	"github.com/Azure/aztfy/internal/client"
 	"github.com/tidwall/gjson"
 )
 
@@ -56,7 +57,7 @@ func populateManagedResources(res Resource, paths ...string) (*Resource, []Resou
 			if !strings.HasPrefix(exprResult.String(), "[") {
 				continue
 			}
-			id, err := NewResourceIdFromCallExpr(exprResult.String())
+			id, err := ParseResourceIdFromCallExpr(exprResult.String())
 			if err != nil {
 				return nil, nil, err
 			}
@@ -78,4 +79,18 @@ func populateManagedResources(res Resource, paths ...string) (*Resource, []Resou
 		}
 	}
 	return &res, resources, nil
+}
+
+// ProviderId converts the ARM ResourceId to its ARM resource ID literal, based on the specified subscription id and resource
+// group name. Then it will optionally transform the ARM resource ID to its corresponding TF resource ID, which might requires
+// API interaction with ARM, where a non-nil client builder is required.
+func (res ResourceId) ProviderId(sub, rg string, b *client.ClientBuilder) (string, error) {
+	switch res.Type {
+	// See issue: https://github.com/Azure/aztfy/issues/89
+	case "microsoft.insights/webtests":
+		res.Type = "Microsoft.insights/webTests"
+		return res.ID(sub, rg), nil
+	default:
+		return res.ID(sub, rg), nil
+	}
 }

--- a/internal/armtemplate/armtemplate_test.go
+++ b/internal/armtemplate/armtemplate_test.go
@@ -184,11 +184,15 @@ func TestUnmarshalTemplate(t *testing.T) {
 func TestDependencyInfo(t *testing.T) {
 	cases := []struct {
 		name   string
+		subId  string
+		rg     string
 		input  armtemplate.Template
-		expect armtemplate.DependencyInfo
+		expect map[string][]string
 	}{
 		{
-			name: "multiple-level dependency",
+			name:  "multiple-level dependency",
+			subId: "sub1",
+			rg:    "rg1",
 			input: armtemplate.Template{
 				Resources: []armtemplate.Resource{
 					{
@@ -259,139 +263,34 @@ func TestDependencyInfo(t *testing.T) {
 					},
 				},
 			},
-			expect: map[armtemplate.ResourceId][]armtemplate.ResourceId{
-				{
-					Type: "Microsoft.Network/networkInterfaces",
-					Name: "nic",
-				}: {
-					{
-						Type: "Microsoft.Network/publicIPAddresses",
-						Name: "pip",
-					},
-					{
-						Type: "Microsoft.Network/virtualNetworks/subnets",
-						Name: "vnet/subnet",
-					},
-					{
-						Type: "Microsoft.Network/networkSecurityGroups",
-						Name: "nsg",
-					},
+			expect: map[string][]string{
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkInterfaces/nic": {
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/publicIPAddresses/pip",
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet",
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/nsg",
 				},
-				{
-					Type: "Microsoft.Network/virtualNetworks/subnets",
-					Name: "vnet/subnet",
-				}: {
-					{
-						Type: "Microsoft.Network/virtualNetworks",
-						Name: "vnet",
-					},
-					{
-						Type: "Microsoft.Network/networkSecurityGroups",
-						Name: "nsg",
-					},
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet/subnets/subnet": {
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet",
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/nsg",
 				},
-				{
-					Type: "Microsoft.Network/networkSecurityGroups/securityRules",
-					Name: "nsg/nsr",
-				}: {
-					{
-						Type: "Microsoft.Network/networkSecurityGroups",
-						Name: "nsg",
-					},
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/nsg/securityRules/nsr": {
+					"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/nsg",
 				},
-				{
-					Type: "Microsoft.Network/networkSecurityGroups",
-					Name: "nsg",
-				}: {
-					armtemplate.ResourceGroupId,
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/nsg": {
+					"/subscriptions/sub1/resourceGroups/rg1",
 				},
-				{
-					Type: "Microsoft.Network/virtualNetworks",
-					Name: "vnet",
-				}: {
-					armtemplate.ResourceGroupId,
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet": {
+					"/subscriptions/sub1/resourceGroups/rg1",
 				},
-				{
-					Type: "Microsoft.Network/publicIPAddresses",
-					Name: "pip",
-				}: {
-					armtemplate.ResourceGroupId,
+				"/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Network/publicIPAddresses/pip": {
+					"/subscriptions/sub1/resourceGroups/rg1",
 				},
 			},
 		},
 	}
 
 	for _, c := range cases {
-		require.Equal(t, c.expect, c.input.DependencyInfo(), c.name)
-	}
-}
-
-func TestNewResourceId(t *testing.T) {
-	cases := []struct {
-		name   string
-		input  string
-		expect armtemplate.ResourceId
-		error  bool
-	}{
-		{
-			name:  "empty",
-			input: "",
-			error: true,
-		},
-		{
-			name:  "only subscription",
-			input: "/subscriptions/1234",
-			error: true,
-		},
-		{
-			name:   "only subscription and resource group",
-			input:  "/subscriptions/1234/resourceGroups/rg1",
-			expect: armtemplate.ResourceGroupId,
-		},
-		{
-			name:  "only subscription, resource group and provider",
-			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network",
-			error: true,
-		},
-		{
-			name:  "valid vnet id",
-			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1",
-			expect: armtemplate.ResourceId{
-				Type: "Microsoft.Network/virtualNetworks",
-				Name: "vnet1",
-			},
-		},
-		{
-			name:  "valid vnet id (small case resourcegroups)",
-			input: "/subscriptions/1234/resourcegroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1",
-			expect: armtemplate.ResourceId{
-				Type: "Microsoft.Network/virtualNetworks",
-				Name: "vnet1",
-			},
-		},
-		{
-			name:  "invalid subnet id",
-			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets",
-			error: true,
-		},
-		{
-			name:  "valid subnet id",
-			input: "/subscriptions/1234/resourcegroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1",
-			expect: armtemplate.ResourceId{
-				Type: "Microsoft.Network/virtualNetworks/subnets",
-				Name: "vnet1/subnet1",
-			},
-		},
-	}
-
-	for _, c := range cases {
-		output, err := armtemplate.NewResourceId(c.input)
-		if c.error {
-			require.Error(t, err, c.name)
-			continue
-		}
-		require.NoError(t, err, c.name)
-		require.Equal(t, c.expect, *output, c.name)
+		require.Equal(t, c.expect, c.input.Qualify(c.subId, c.rg).DependencyInfo(), c.name)
 	}
 }
 

--- a/internal/armtemplate/armtemplate_test.go
+++ b/internal/armtemplate/armtemplate_test.go
@@ -8,6 +8,153 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestParseResourceId(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  string
+		expect armtemplate.ResourceId
+		error  bool
+	}{
+		{
+			name:  "empty",
+			input: "",
+			error: true,
+		},
+		{
+			name:  "only subscription",
+			input: "/subscriptions/1234",
+			error: true,
+		},
+		{
+			name:   "only subscription and resource group",
+			input:  "/subscriptions/1234/resourceGroups/rg1",
+			expect: armtemplate.ResourceGroupId,
+		},
+		{
+			name:  "only subscription, resource group and provider",
+			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network",
+			error: true,
+		},
+		{
+			name:  "valid vnet id",
+			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1",
+			expect: armtemplate.ResourceId{
+				Type: "Microsoft.Network/virtualNetworks",
+				Name: "vnet1",
+			},
+		},
+		{
+			name:  "valid vnet id (small case resourcegroups)",
+			input: "/subscriptions/1234/resourcegroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1",
+			expect: armtemplate.ResourceId{
+				Type: "Microsoft.Network/virtualNetworks",
+				Name: "vnet1",
+			},
+		},
+		{
+			name:  "invalid subnet id",
+			input: "/subscriptions/1234/resourceGroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets",
+			error: true,
+		},
+		{
+			name:  "valid subnet id",
+			input: "/subscriptions/1234/resourcegroups/rg1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1",
+			expect: armtemplate.ResourceId{
+				Type: "Microsoft.Network/virtualNetworks/subnets",
+				Name: "vnet1/subnet1",
+			},
+		},
+	}
+
+	for _, c := range cases {
+		output, err := armtemplate.ParseResourceId(c.input)
+		if c.error {
+			require.Error(t, err, c.name)
+			continue
+		}
+		require.NoError(t, err, c.name)
+		require.Equal(t, c.expect, *output, c.name)
+	}
+}
+
+func TestParseResourceIdFromCallExpr(t *testing.T) {
+	cases := []struct {
+		name   string
+		expr   string
+		expect armtemplate.ResourceId
+		error  bool
+	}{
+		{
+			name:  "empty",
+			expr:  "",
+			error: true,
+		},
+		{
+			name:  "no args",
+			expr:  "[resourceId()]",
+			error: true,
+		},
+		{
+			name: "one level",
+			expr: "[resourceId('Microsoft.Storage/storageAccounts', 'a')]",
+			expect: armtemplate.ResourceId{
+				Type: "Microsoft.Storage/storageAccounts",
+				Name: "a",
+			},
+			error: false,
+		},
+		{
+			name: "two levels",
+			expr: "[resourceId('Microsoft.Storage/storageAccounts/services', 'a', 'b')]",
+			expect: armtemplate.ResourceId{
+				Type: "Microsoft.Storage/storageAccounts/services",
+				Name: "a/b",
+			},
+			error: false,
+		},
+	}
+
+	for _, c := range cases {
+		output, err := armtemplate.ParseResourceIdFromCallExpr(c.expr)
+		if c.error {
+			require.Error(t, err, c.name)
+			continue
+		}
+		require.NoError(t, err, c.name)
+		require.Equal(t, c.expect, *output, c.name)
+	}
+}
+
+func TestResourceId_ID(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  armtemplate.ResourceId
+		expect string
+	}{
+		{
+			name: "one level",
+			input: armtemplate.ResourceId{
+				Type: "Microsoft.Storage/storageAccounts",
+				Name: "a",
+			},
+			expect: "/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Storage/storageAccounts/a",
+		},
+		{
+			name: "two levels",
+			input: armtemplate.ResourceId{
+				Type: "Microsoft.Storage/storageAccounts/services",
+				Name: "a/b",
+			},
+			expect: "/subscriptions/sub1/resourceGroups/rg1/providers/Microsoft.Storage/storageAccounts/a/services/b",
+		},
+	}
+
+	for _, c := range cases {
+		actual := c.input.ID("sub1", "rg1")
+		require.Equal(t, actual, c.expect, c.name)
+	}
+}
+
 func TestUnmarshalTemplate(t *testing.T) {
 	cases := []struct {
 		name   string
@@ -290,54 +437,10 @@ func TestDependencyInfo(t *testing.T) {
 	}
 
 	for _, c := range cases {
-		require.Equal(t, c.expect, c.input.Qualify(c.subId, c.rg).DependencyInfo(), c.name)
-	}
-}
-
-func TestNewResourceIdFromCallExpr(t *testing.T) {
-	cases := []struct {
-		name   string
-		expr   string
-		expect armtemplate.ResourceId
-		error  bool
-	}{
-		{
-			name:  "empty",
-			expr:  "",
-			error: true,
-		},
-		{
-			name:  "no args",
-			expr:  "[resourceId()]",
-			error: true,
-		},
-		{
-			name: "one level",
-			expr: "[resourceId('Microsoft.Storage/storageAccounts', 'a')]",
-			expect: armtemplate.ResourceId{
-				Type: "Microsoft.Storage/storageAccounts",
-				Name: "a",
-			},
-			error: false,
-		},
-		{
-			name: "two levels",
-			expr: "[resourceId('Microsoft.Storage/storageAccounts/services', 'a', 'b')]",
-			expect: armtemplate.ResourceId{
-				Type: "Microsoft.Storage/storageAccounts/services",
-				Name: "a/b",
-			},
-			error: false,
-		},
-	}
-
-	for _, c := range cases {
-		output, err := armtemplate.NewResourceIdFromCallExpr(c.expr)
-		if c.error {
-			require.Error(t, err, c.name)
-			continue
-		}
+		tpl, err := c.input.Qualify(c.subId, c.rg, nil)
 		require.NoError(t, err, c.name)
-		require.Equal(t, c.expect, *output, c.name)
+		deps, _ := tpl.DependencyInfo()
+		require.NoError(t, err, c.name)
+		require.Equal(t, c.expect, deps, c.name)
 	}
 }

--- a/internal/client/client.go
+++ b/internal/client/client.go
@@ -1,4 +1,4 @@
-package meta
+package client
 
 import (
 	"fmt"

--- a/internal/meta/hcl_edit.go
+++ b/internal/meta/hcl_edit.go
@@ -1,0 +1,57 @@
+package meta
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+)
+
+func hclBlockAppendDependency(body *hclwrite.Body, ids []string, cfgset map[string]ConfigInfo) error {
+	dependencies := []string{}
+	for _, id := range ids {
+		cfg, ok := cfgset[id]
+		if !ok {
+			dependencies = append(dependencies, fmt.Sprintf("# Depending on %q, which is not imported by Terraform.", id))
+			continue
+		}
+		dependencies = append(dependencies, cfg.TFAddr.String()+",")
+	}
+	if len(dependencies) > 0 {
+		src := []byte("depends_on = [\n" + strings.Join(dependencies, "\n") + "\n]")
+		expr, diags := hclwrite.ParseConfig(src, "f", hcl.InitialPos)
+		if diags.HasErrors() {
+			return fmt.Errorf(`building "depends_on" attribute: %s`, diags.Error())
+		}
+
+		body.SetAttributeRaw("depends_on", expr.Body().GetAttribute("depends_on").Expr().BuildTokens(nil))
+	}
+
+	return nil
+}
+
+func hclBlockAppendLifecycle(body *hclwrite.Body, ignoreChanges []string) error {
+	srcs := map[string][]byte{}
+	if len(ignoreChanges) > 0 {
+		for i := range ignoreChanges {
+			ignoreChanges[i] = ignoreChanges[i] + ","
+		}
+		srcs["ignore_changes"] = []byte("ignore_changes = [\n" + strings.Join(ignoreChanges, "\n") + "\n]\n")
+	}
+
+	if len(srcs) == 0 {
+		return nil
+	}
+
+	b := hclwrite.NewBlock("lifecycle", nil)
+	for name, src := range srcs {
+		expr, diags := hclwrite.ParseConfig(src, "f", hcl.InitialPos)
+		if diags.HasErrors() {
+			return fmt.Errorf(`building "lifecycle.%s" attribute: %s`, name, diags.Error())
+		}
+		b.Body().SetAttributeRaw(name, expr.Body().GetAttribute(name).Expr().BuildTokens(nil))
+	}
+	body.AppendBlock(b)
+	return nil
+}

--- a/internal/meta/hcl_edit_test.go
+++ b/internal/meta/hcl_edit_test.go
@@ -1,0 +1,39 @@
+package meta
+
+import (
+	"testing"
+
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHclBlockAppendLifecycle(t *testing.T) {
+	cases := []struct {
+		name          string
+		ignoreChanges []string
+		expect        string
+	}{
+		{
+			name:          "no lifecycle should be generated",
+			ignoreChanges: nil,
+			expect:        "",
+		},
+		{
+			name:          "with ignore_changes",
+			ignoreChanges: []string{"foo", "bar"},
+			expect: `lifecycle {
+  ignore_changes = [
+    foo,
+    bar,
+  ]
+}
+`,
+		},
+	}
+
+	for _, c := range cases {
+		b := hclwrite.NewFile().Body()
+		require.NoError(t, hclBlockAppendLifecycle(b, c.ignoreChanges), c.name)
+		require.Equal(t, string(hclwrite.Format(b.BuildTokens(nil).Bytes())), c.expect, c.name)
+	}
+}

--- a/internal/meta/meta_impl.go
+++ b/internal/meta/meta_impl.go
@@ -11,10 +11,12 @@ import (
 	"strings"
 	"time"
 
+	"github.com/Azure/aztfy/internal/client"
+
+	"github.com/Azure/aztfy/internal/armtemplate"
 	"github.com/Azure/aztfy/internal/resmap"
 	"github.com/Azure/aztfy/internal/tfaddr"
 
-	"github.com/Azure/aztfy/internal/armtemplate"
 	"github.com/Azure/aztfy/internal/config"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
 
@@ -31,7 +33,7 @@ type MetaImpl struct {
 	rootdir        string
 	outdir         string
 	tf             *tfexec.Terraform
-	clientBuilder  *ClientBuilder
+	clientBuilder  *client.ClientBuilder
 	armTemplate    armtemplate.FQTemplate
 
 	// Key is azure resource id; Value is terraform resource addr.
@@ -106,7 +108,7 @@ func newMetaImpl(cfg config.Config) (Meta, error) {
 	}
 
 	// Construct client builder
-	b, err := NewClientBuilder()
+	b, err := client.NewClientBuilder()
 	if err != nil {
 		return nil, fmt.Errorf("building authorizer: %w", err)
 	}
@@ -178,7 +180,8 @@ func (meta *MetaImpl) ListResource() (ImportList, error) {
 	for _, res := range meta.armTemplate.Resources {
 		ids = append(ids, res.Id)
 	}
-	ids = append(ids, armtemplate.ResourceGroupId.ID(meta.subscriptionId, meta.resourceGroup))
+	rgid, _ := armtemplate.ResourceGroupId.ProviderId(meta.subscriptionId, meta.resourceGroup, nil)
+	ids = append(ids, rgid)
 
 	l := make(ImportList, 0, len(ids))
 	for i, id := range ids {
@@ -233,10 +236,11 @@ func (meta MetaImpl) GenerateCfg(l ImportList) error {
 	if err != nil {
 		return fmt.Errorf("converting from state to configurations: %w", err)
 	}
-	cfginfos, err = meta.resolveDependency(cfginfos)
+	cfginfos, err = meta.terraformMetaHook(cfginfos)
 	if err != nil {
-		return fmt.Errorf("resolving cross resource dependencies: %w", err)
+		return fmt.Errorf("Terraform HCL meta hook: %w", err)
 	}
+
 	return meta.generateConfig(cfginfos)
 }
 
@@ -329,7 +333,12 @@ func (meta *MetaImpl) exportArmTemplate(ctx context.Context) error {
 	if err := tpl.PopulateManagedResources(); err != nil {
 		return fmt.Errorf("populating managed resources in the ARM template: %v", err)
 	}
-	meta.armTemplate = tpl.Qualify(meta.subscriptionId, meta.resourceGroup)
+
+	fqTpl, err := tpl.Qualify(meta.subscriptionId, meta.resourceGroup, meta.clientBuilder)
+	if err != nil {
+		return err
+	}
+	meta.armTemplate = *fqTpl
 
 	return nil
 }
@@ -356,8 +365,24 @@ func (meta MetaImpl) stateToConfig(ctx context.Context, list ImportList) (Config
 	return out, nil
 }
 
+func (meta MetaImpl) terraformMetaHook(configs ConfigInfos) (ConfigInfos, error) {
+	var err error
+	configs, err = meta.resolveDependency(configs)
+	if err != nil {
+		return nil, fmt.Errorf("resolving cross resource dependencies: %w", err)
+	}
+	configs, err = meta.lifecycleAddon(configs)
+	if err != nil {
+		return nil, fmt.Errorf("adding terraform lifecycle: %w", err)
+	}
+	return configs, nil
+}
+
 func (meta MetaImpl) resolveDependency(configs ConfigInfos) (ConfigInfos, error) {
-	depInfo := meta.armTemplate.DependencyInfo()
+	depInfo, err := meta.armTemplate.DependencyInfo()
+	if err != nil {
+		return nil, err
+	}
 
 	configSet := map[string]ConfigInfo{}
 	for _, cfg := range configs {
@@ -366,8 +391,9 @@ func (meta MetaImpl) resolveDependency(configs ConfigInfos) (ConfigInfos, error)
 
 	// Iterate each config to add dependency by querying the dependency info from arm template.
 	var out ConfigInfos
+	rgid, _ := armtemplate.ResourceGroupId.ProviderId(meta.subscriptionId, meta.resourceGroup, nil)
 	for id, cfg := range configSet {
-		if id == armtemplate.ResourceGroupId.ID(meta.subscriptionId, meta.resourceGroup) {
+		if id == rgid {
 			out = append(out, cfg)
 			continue
 		}
@@ -376,36 +402,13 @@ func (meta MetaImpl) resolveDependency(configs ConfigInfos) (ConfigInfos, error)
 			return nil, fmt.Errorf("can't find resource %q in the arm template", id)
 		}
 
-		if err := meta.hclBlockAppendDependency(cfg.hcl.Body().Blocks()[0].Body(), depInfo[id], configSet); err != nil {
+		if err := hclBlockAppendDependency(cfg.hcl.Body().Blocks()[0].Body(), depInfo[id], configSet); err != nil {
 			return nil, err
 		}
 		out = append(out, cfg)
 	}
 
 	return out, nil
-}
-
-func (meta MetaImpl) hclBlockAppendDependency(body *hclwrite.Body, ids []string, cfgset map[string]ConfigInfo) error {
-	dependencies := []string{}
-	for _, id := range ids {
-		cfg, ok := cfgset[id]
-		if !ok {
-			dependencies = append(dependencies, fmt.Sprintf("# Depending on %q, which is not imported by Terraform.", id))
-			continue
-		}
-		dependencies = append(dependencies, cfg.TFAddr.String()+",")
-	}
-	if len(dependencies) > 0 {
-		src := []byte("depends_on = [\n" + strings.Join(dependencies, "\n") + "\n]")
-		expr, diags := hclwrite.ParseConfig(src, "generate_depends_on", hcl.InitialPos)
-		if diags.HasErrors() {
-			return fmt.Errorf(`building "depends_on" attribute: %s`, diags.Error())
-		}
-
-		body.SetAttributeRaw("depends_on", expr.Body().GetAttribute("depends_on").Expr().BuildTokens(nil))
-	}
-
-	return nil
 }
 
 func (meta MetaImpl) generateConfig(cfgs ConfigInfos) error {

--- a/internal/meta/meta_impl.go
+++ b/internal/meta/meta_impl.go
@@ -32,7 +32,7 @@ type MetaImpl struct {
 	outdir         string
 	tf             *tfexec.Terraform
 	clientBuilder  *ClientBuilder
-	armTemplate    armtemplate.Template
+	armTemplate    armtemplate.FQTemplate
 
 	// Key is azure resource id; Value is terraform resource addr.
 	// For azure resources not in this mapping, they are all initialized as to skip.
@@ -176,7 +176,7 @@ func (meta *MetaImpl) ListResource() (ImportList, error) {
 
 	var ids []string
 	for _, res := range meta.armTemplate.Resources {
-		ids = append(ids, res.ID(meta.subscriptionId, meta.resourceGroup))
+		ids = append(ids, res.Id)
 	}
 	ids = append(ids, armtemplate.ResourceGroupId.ID(meta.subscriptionId, meta.resourceGroup))
 
@@ -322,13 +322,14 @@ func (meta *MetaImpl) exportArmTemplate(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("marshalling the template: %w", err)
 	}
-	if err := json.Unmarshal(raw, &meta.armTemplate); err != nil {
+	var tpl armtemplate.Template
+	if err := json.Unmarshal(raw, &tpl); err != nil {
 		return fmt.Errorf("unmarshalling the template: %w", err)
 	}
-
-	if err := meta.armTemplate.PopulateManagedResources(); err != nil {
+	if err := tpl.PopulateManagedResources(); err != nil {
 		return fmt.Errorf("populating managed resources in the ARM template: %v", err)
 	}
+	meta.armTemplate = tpl.Qualify(meta.subscriptionId, meta.resourceGroup)
 
 	return nil
 }
@@ -358,28 +359,24 @@ func (meta MetaImpl) stateToConfig(ctx context.Context, list ImportList) (Config
 func (meta MetaImpl) resolveDependency(configs ConfigInfos) (ConfigInfos, error) {
 	depInfo := meta.armTemplate.DependencyInfo()
 
-	configSet := map[armtemplate.ResourceId]ConfigInfo{}
+	configSet := map[string]ConfigInfo{}
 	for _, cfg := range configs {
-		armId, err := armtemplate.NewResourceId(cfg.ResourceID)
-		if err != nil {
-			return nil, fmt.Errorf("new arm tempalte resource id from azure resource id: %w", err)
-		}
-		configSet[*armId] = cfg
+		configSet[cfg.ResourceID] = cfg
 	}
 
 	// Iterate each config to add dependency by querying the dependency info from arm template.
 	var out ConfigInfos
-	for armId, cfg := range configSet {
-		if armId == armtemplate.ResourceGroupId {
+	for id, cfg := range configSet {
+		if id == armtemplate.ResourceGroupId.ID(meta.subscriptionId, meta.resourceGroup) {
 			out = append(out, cfg)
 			continue
 		}
 		// This should never happen as we always ensure there is at least one implicit dependency on the resource group for each resource.
-		if _, ok := depInfo[armId]; !ok {
-			return nil, fmt.Errorf("can't find resource %q in the arm template", armId.ID(meta.subscriptionId, meta.resourceGroup))
+		if _, ok := depInfo[id]; !ok {
+			return nil, fmt.Errorf("can't find resource %q in the arm template", id)
 		}
 
-		if err := meta.hclBlockAppendDependency(cfg.hcl.Body().Blocks()[0].Body(), depInfo[armId], configSet); err != nil {
+		if err := meta.hclBlockAppendDependency(cfg.hcl.Body().Blocks()[0].Body(), depInfo[id], configSet); err != nil {
 			return nil, err
 		}
 		out = append(out, cfg)
@@ -388,12 +385,12 @@ func (meta MetaImpl) resolveDependency(configs ConfigInfos) (ConfigInfos, error)
 	return out, nil
 }
 
-func (meta MetaImpl) hclBlockAppendDependency(body *hclwrite.Body, armIds []armtemplate.ResourceId, cfgset map[armtemplate.ResourceId]ConfigInfo) error {
+func (meta MetaImpl) hclBlockAppendDependency(body *hclwrite.Body, ids []string, cfgset map[string]ConfigInfo) error {
 	dependencies := []string{}
-	for _, armid := range armIds {
-		cfg, ok := cfgset[armid]
+	for _, id := range ids {
+		cfg, ok := cfgset[id]
 		if !ok {
-			dependencies = append(dependencies, fmt.Sprintf("# Depending on %q, which is not imported by Terraform.", armid.ID(meta.subscriptionId, meta.resourceGroup)))
+			dependencies = append(dependencies, fmt.Sprintf("# Depending on %q, which is not imported by Terraform.", id))
 			continue
 		}
 		dependencies = append(dependencies, cfg.TFAddr.String()+",")

--- a/internal/meta/terraform_hack.go
+++ b/internal/meta/terraform_hack.go
@@ -1,0 +1,47 @@
+package meta
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+)
+
+// lifecycleAddon adds lifecycle meta arguments for some identified resources, which are mandatory to make them usable.
+func (meta MetaImpl) lifecycleAddon(configs ConfigInfos) (ConfigInfos, error) {
+	out := make(ConfigInfos, len(configs))
+	for i, cfg := range configs {
+		switch cfg.TFAddr.Type {
+		case "azurerm_application_insights_web_test":
+			if err := hclBlockAppendLifecycle(cfg.hcl.Body().Blocks()[0].Body(), []string{"tags"}); err != nil {
+				return nil, fmt.Errorf("azurerm_application_insights_web_test: %v", err)
+			}
+		}
+		out[i] = cfg
+	}
+	return out, nil
+}
+
+func (meta MetaImpl) hclBlockAppendDependency(body *hclwrite.Body, ids []string, cfgset map[string]ConfigInfo) error {
+	dependencies := []string{}
+	for _, id := range ids {
+		cfg, ok := cfgset[id]
+		if !ok {
+			dependencies = append(dependencies, fmt.Sprintf("# Depending on %q, which is not imported by Terraform.", id))
+			continue
+		}
+		dependencies = append(dependencies, cfg.TFAddr.String()+",")
+	}
+	if len(dependencies) > 0 {
+		src := []byte("depends_on = [\n" + strings.Join(dependencies, "\n") + "\n]")
+		expr, diags := hclwrite.ParseConfig(src, "generate_depends_on", hcl.InitialPos)
+		if diags.HasErrors() {
+			return fmt.Errorf(`building "depends_on" attribute: %s`, diags.Error())
+		}
+
+		body.SetAttributeRaw("depends_on", expr.Body().GetAttribute("depends_on").Expr().BuildTokens(nil))
+	}
+
+	return nil
+}

--- a/internal/test/case_applicationinsight_webtest.go
+++ b/internal/test/case_applicationinsight_webtest.go
@@ -1,0 +1,60 @@
+package test
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Azure/aztfy/internal/resmap"
+)
+
+type CaseApplicationInsightWebTest struct{}
+
+func (CaseApplicationInsightWebTest) Tpl(d Data) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+resource "azurerm_resource_group" "test" {
+  name     = "%[1]s"
+  location = "WestEurope"
+}
+resource "azurerm_application_insights" "test" {
+  name                = "test-%[2]s"
+  location            = azurerm_resource_group.test.location
+  resource_group_name = azurerm_resource_group.test.name
+  application_type    = "web"
+}
+resource "azurerm_application_insights_web_test" "test" {
+  name                    = "test-%[2]s"
+  location                = azurerm_resource_group.test.location
+  resource_group_name     = azurerm_resource_group.test.name
+  application_insights_id = azurerm_application_insights.test.id
+  kind                    = "ping"
+  geo_locations           = ["us-tx-sn1-azr"]
+  configuration           = <<XML
+<WebTest Name="WebTest1" Id="ABD48585-0831-40CB-9069-682EA6BB3583" Enabled="True" CssProjectStructure="" CssIteration="" Timeout="0" WorkItemIds="" xmlns="http://microsoft.com/schemas/VisualStudio/TeamTest/2010" Description="" CredentialUserName="" CredentialPassword="" PreAuthenticate="True" Proxy="default" StopOnError="False" RecordedResultFile="" ResultsLocale="">
+  <Items>
+    <Request Method="GET" Guid="a5f10126-e4cd-570d-961c-cea43999a200" Version="1.1" Url="http://microsoft.com" ThinkTime="0" Timeout="300" ParseDependentRequests="True" FollowRedirects="True" RecordResult="True" Cache="False" ResponseTimeGoal="0" Encoding="utf-8" ExpectedHttpStatusCode="200" ExpectedResponseUrl="" ReportingName="" IgnoreHttpStatusCode="False" />
+  </Items>
+</WebTest>
+XML
+  lifecycle {
+    ignore_changes = ["tags"]
+  }
+}
+`, d.RandomRgName(), d.RandomStringOfLength(8))
+}
+
+func (CaseApplicationInsightWebTest) ResourceMapping(d Data) resmap.ResourceMapping {
+	rm := fmt.Sprintf(`{
+"/subscriptions/%[1]s/resourceGroups/%[2]s": "azurerm_resource_group.test",
+"/subscriptions/%[1]s/resourceGroups/%[2]s/providers/microsoft.insights/components/test-%[3]s": "azurerm_application_insights.test",
+"/subscriptions/%[1]s/resourceGroups/%[2]s/providers/Microsoft.insights/webTests/test-%[3]s": "azurerm_application_insights_web_test.test"
+}
+`, d.subscriptionId, d.RandomRgName(), d.RandomStringOfLength(8))
+	m := resmap.ResourceMapping{}
+	if err := json.Unmarshal([]byte(rm), &m); err != nil {
+		panic(err)
+	}
+	return m
+}

--- a/internal/test/e2e_test.go
+++ b/internal/test/e2e_test.go
@@ -110,3 +110,10 @@ func TestComputeVMDisk(t *testing.T) {
 	c, d := CaseComputeVMDisk{}, NewData()
 	runCase(t, d, c)
 }
+
+func TestApplicationInsightWebTest(t *testing.T) {
+	t.Parallel()
+	precheck(t)
+	c, d := CaseApplicationInsightWebTest{}, NewData()
+	runCase(t, d, c)
+}

